pallet-parachain-upgrade/src/lib.rs

```diff
@@ -0,0 +1,559 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Encode;
+use cumulus_runtime::validation_function_params::{
+	ValidationFunctionParams,
+	NEW_VALIDATION_CODE, VALIDATION_FUNCTION_PARAMS, INHERENT_IDENTIFIER,
+};
+use frame_support::{
+	decl_error, decl_event, decl_module, decl_storage, ensure, storage, traits::Get,
+};
+use parachain::primitives::RelayChainBlockNumber;
+use sp_core::storage::well_known_keys;
+use sp_inherents::{ProvideInherent, InherentData, InherentIdentifier};
+use sp_version::RuntimeVersion;
+
+pub type ValidationFunction = Vec<u8>;
+type System<T> = system::Module<T>;
+
+/// The pallet's configuration trait.
+pub trait Trait: system::Trait {
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+
+	/// Get the chain's current version.
+	type Version: Get<RuntimeVersion>;
+}
+
+// This pallet's storage items.
+decl_storage! {
+	trait Store for Module<T: Trait> as ParachainUpgrade {
+		// we need to store the new validation function for the span between
+		// setting it and applying it.
+		PendingValidationFunction get(fn new_validation_function): (RelayChainBlockNumber, ValidationFunction);
+
+		// we keep the validation function parameters here because we have to use
+		// storage as a transport layer between the inherents and the module
+		VFPs get(fn vfps): ValidationFunctionParams;
+	}
+}
+
+// The pallet's dispatchable functions.
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		// Initializing events
+		// this is needed only if you are using events in your pallet
+		fn deposit_event() = default;
+
+		fn on_initialize(_n: T::BlockNumber) {
+			if let Ok((apply_block, validation_function)) = PendingValidationFunction::try_get() {
```

Comment:
Please take a look at the `frame-scheduler` for doing this stuff.

Reply:
I see it, and it's cool, but it's not obvious to me what action you intend me to take from this suggestion. Do you mean that I should build on that pallet for implementing this functionality? That kind of feels like overkill, to me.

Reply:
Why is that an overkill?

Reply:
This pallet exists exactly for this use case.

Reply:
- Storage: implementing the delayed update ourselves requires storing a tuple of `(u32, Vec<u8>)`. At the minimum, storing it in the scheduler requires storing a map of `u32 => Option<Vec<dyn Fn>>`, which likely takes more storage space.
- Performance: on every block, the schedule pallet has to load the appropriate vector from the storage map and iterate through it. I believe that this is slower than just checking for existence, which is the only check I make when doing it myself. 
- Implementation. The code which exists in this pallet has some tests and appears to work. Rewriting it in terms of the scheduler would take additional time, digging through resources, etc. While I'm willing to bet that there are examples to find of other pallets using the schedule pallet, working from example is slower than working from documentation, which is scarce.

I might be wrong about any or all of those points; I'd be glad to learn so, if that's the case. However, there's one point more important than those: the schedule pallet operates in terms of the chain's native block number: the parachain block number. That doesn't work for our use case: we check the planned upgrade against the relay chain block number, and ignore the parachain block number. Looking at its code, I can't see a way to make the schedule pallet do what we want in that respect without significant modification.